micro-rdk/src/esp32/coredump.rs

```diff
@@ -0,0 +1,196 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use base64::{engine::general_purpose::STANDARD, Engine as _};
+use bytes::BytesMut;
+use esp_idf_svc::sys::{
+    esp_partition_erase_range, esp_partition_find_first, esp_partition_read,
+    esp_partition_subtype_t_ESP_PARTITION_SUBTYPE_DATA_COREDUMP, esp_partition_t,
+    esp_partition_type_t_ESP_PARTITION_TYPE_DATA,
+};
+
+use crate::{
+    common::{
+        config::ConfigType,
+        generic::{DoCommand, GenericError},
+        registry::{ComponentRegistry, Dependency},
+        sensor::{GenericReadingsResult, Readings, Sensor, SensorError, SensorType},
+        status::Status,
+    },
+    google::protobuf::{self, value::Kind, Struct, Value},
+};
+
+pub struct Coredump {
+    len: usize,
+    coredump_partition_ptr: &'static esp_partition_t,
+}
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_sensor("coredump", &Coredump::from_config)
+        .is_err()
+    {
+        log::error!("couldn't register coredump sensor");
+    }
+}
+
+/// Instantiate a coredump sensor if : any partition of type data and subtype coredump can be found, we will use the first found partition is more than one exists (similar to the behavior of the coredump code in esp-idf)
+/// A coredump is consider available if the 4th byte of the version field is not 0xFF (flash erase value)
+/// The coredump is implemented as a sensor and can be downloaded through the DoCommand interface ( since GetReading cannot return a stream of bytes object)
+impl Coredump {
+    pub fn from_config(_: ConfigType, _: Vec<Dependency>) -> Result<SensorType, SensorError> {
+        let mut len = 0;
+        // if a pointer is found it will be stored as a static reference since it will remain valid for the lifetime of the program
+        let partition_iterator = unsafe {
+            esp_partition_find_first(
+                esp_partition_type_t_ESP_PARTITION_TYPE_DATA,
+                esp_partition_subtype_t_ESP_PARTITION_SUBTYPE_DATA_COREDUMP,
+                std::ptr::null(),
+            )
+            .as_ref()
+            .ok_or(SensorError::SensorGenericError(
+                "no coredump partition found",
+            ))?
+        };
+        // the first two fields of the coredump are length (uint32_t) and version (uint32_t)
+        let mut first_bytes = [0xFF_u8; 8];
+        esp_idf_svc::sys::esp!(unsafe {
+            esp_partition_read(partition_iterator, 0, first_bytes.as_mut_ptr() as *mut _, 8)
+        })?;
+
+        if first_bytes[7] != 0xFF {
+            len = usize::from_le_bytes(first_bytes[0..4].try_into().unwrap()); // safe because we would have read 4 bytes
```

Comment:
just to make sure I understand: we only need to do this once because even if someone erases the coredump via DoCommand, a coredump can only be written in a situation that what prompt a restart and result in reconfiguration? Is this correct? Or is there a situation in which something can be written to the coredump and a restart does not occur (meaning this sensor's state won't be properly reset)?

Reply:
coredump is only written on panic/abort whether from rust or C. So during the lifetime of the program the coredump exists until we erase it